design/accepted/gc_plus/uncomitted-gc.md

```diff
@@ -0,0 +1,203 @@
+quickly read all uncommitted objects in the repository.
+
+### Required changes by lakeFS
+
+The following are necessary changes in lakeFS in order to implement this proposal successfully.
+
+#### Objects Path Conventions
+
+Uncommitted GC must scan the bucket in order to find objects that are not referenced by lakeFS.
+To optimize this process, suggest the following changes:
+
+1. Store lakeFS data under `<namespace>/data/` path
+2. Divide the repository data path into time-and-size based slices.
+3. Slice name will be a time based, reverse sorted unique identifier.
+4. LakeFS will create a new slice on a timely basis (for example: hourly) or when it has written < MAX_SLICE_SIZE > objects to the slice.
+5. Each slice will be written by a single lakeFS instance in order to track slice size.
+6. The sorted slices will enable partial scans of the bucket when running the optimized GC.
+
+#### StageObject
+
+The StageObject operation will only be allowed on addresses outside the repository's storage namespace. This way, 
+objects added using this operation are never collected by GC.
+
+#### [Get/Link]PhysicalAddress 
+
+1. GetPhysicalAddress to return a validation token along with the address (or embedded as part of the address).
+2. The token will be valid for a specified amount of time and for a single use.
+3. lakeFS will need to track issued tokens/addresses, and delete them when tokens are expired/used
+4. LinkPhysicalAddress to verify token valid before creating an entry.
+    1. Doing so will allow us to use this time interval to filter objects that might have been uploaded and waiting for
+       the link API and avoid them being deleted by the GC process.
+    2. Objects that were uploaded to a physical address issued by the API and were not linked before the token expired will
+       eventually be deleted by the GC job.
+>**Note:** These changes will also solves the following [issue](https://github.com/treeverse/lakeFS/issues/4438)
+
+#### S3 Gateway CopyObject
+
+1. Copy object in the same branch will work the same - creating a new staging entry using the existing entry information.
+2. For objects that are not part of the branch, use the underlying adapter copy operation.
+
+#### Move/RenameObject
+Clients working through the S3 Gateway can use the CopyObject + DeleteObject to perform a Rename or Move operation.
+For clients using the OpenAPI this could have been done using StageObject + DeleteObject.
+To continue support of this operation, introduce a new API to rename an object which will be scoped to a single branch.
+
+#### PrepareUncommittedForGC
+
+A new API which will create meta-ranges and ranges for a given branch using its uncommitted data. These files
+will be saved in a designated path used by the GC client to list branch's uncommitted objects.  
+For the purpose of this document we'll call this the `BranchUncommittedMetarange`
```

Comment:
I think there's a bug.
I start preparing the uncommitted objects. During that time a Copy is happening on the same branch, making it a metadata operation. The copy copies from the path `zz` to the path `aa`. The preparation of the uncommitted started before `aa` was created, but by the time it reaches `zz` it was already deleted. Now I don't see neither of them in the output file and the physical object may be deleted by the GC.

Reply:
Thanks! Addressed in proposal